src/sinks/prometheus/exporter.rs

```diff
@@ -160,39 +179,160 @@ impl SinkConfig for PrometheusCompatConfig {
 struct PrometheusExporter {
     server_shutdown_trigger: Option<Trigger>,
     config: PrometheusExporterConfig,
-    metrics: Arc<RwLock<ExpiringMetrics>>,
+    metrics: Arc<RwLock<IndexMap<MetricRef, (Metric, MetricMetadata)>>>,
     acker: Acker,
 }
 
-struct ExpiringMetrics {
-    map: IndexMap<MetricEntry, MetricMetadata>,
-    last_flush_timestamp: i64,
+/// Expiration metadata for a metric.
+#[derive(Clone, Copy, Debug)]
+struct MetricMetadata {
+    should_clear: bool,
+    last_updated: Instant,
 }
 
-struct MetricMetadata {
-    is_incremental_set: bool,
-    updated_at: Instant,
+impl MetricMetadata {
+    pub fn new(should_clear: bool) -> Self {
+        Self {
+            should_clear,
+            last_updated: Instant::now(),
+        }
+    }
+
+    pub fn has_expired(&self, timeout: Duration) -> bool {
+        self.last_updated.elapsed() > timeout
+    }
+}
+
+// Composite identifier that uniquely represents a metric.
+//
+// Instead of simply working off of the name (series) alone, we include the metric kind as well as
+// the type (counter, gauge, etc) and any subtype information like histogram buckets.
+//
+// Specifically, though, we do _not_ include the actual metric value.  This type is used
+// specifically to look up the entry in a map for a metric in the sense of "get the metric whose
+// name is X and type is Y and has these tags".
+#[derive(Clone, Debug)]
+struct MetricRef(
+    MetricSeries,
+    MetricKind,
+    Discriminant<MetricValue>,
+    Option<Vec<f64>>,
+);
```

Comment:
Using a tuple struct with more than one or two elements can get confusing. I recommend naming the fields here.

Reply:
Yep, I can get behind that sentiment.

I've gone ahead and named struct-ified it.